packages/zoltan2/core/src/models/Zoltan2_CoordinateModel.hpp

```diff
@@ -356,6 +350,25 @@ void CoordinateModel<Adapter>::sharedConstructor(
 
   if (userNumWeights_)
     weights_ = arcp(weightArray, 0, userNumWeights_);
+
+  // These are deep copies so we don't hold on to refs of the device views causing problems without UVM
+  Kokkos::View<const gno_t *, typename node_t::device_type> kokkos_gids;
+  ia->getIDsKokkosView(kokkos_gids);
+  kokkos_gids_ =   Kokkos::View<gno_t *, typename node_t::device_type>("kokkos_gids_",kokkos_gids.extent(0));
+  Kokkos::deep_copy(kokkos_gids_, kokkos_gids);
+
+  Kokkos::View<scalar_t **, Kokkos::LayoutLeft, typename node_t::device_type> kokkos_xyz;
+  ia->getCoordinatesKokkosView(kokkos_xyz);
+  kokkos_xyz_ =   Kokkos::View<scalar_t **, Kokkos::LayoutLeft, typename node_t::device_type>("kokkos_xyz", kokkos_xyz.extent(0), kokkos_xyz.extent(1));
+  Kokkos::deep_copy(kokkos_xyz_, kokkos_xyz);
+
+  if(userNumWeights_ > 0) {
+    Kokkos::View<scalar_t **, typename node_t::device_type> kokkos_weights;
+    ia->getWeightsKokkosView(kokkos_weights);
+    kokkos_weights_ = Kokkos::View<scalar_t **, typename node_t::device_type>("kokkos_weights_",kokkos_weights.extent(0), kokkos_weights.extent(1));
+    Kokkos::deep_copy(kokkos_weights_, kokkos_weights);
+  }
+
```

Comment:
The deep copies are OK for now.  

When we circle back to Zoltan2, I'd like to try avoiding the copies by not saving the pointers in the class at all.  Instead, we could rewrite getCoordinatesKokkos to call the input adapter functions and return the device views without storing them in the class (e.g., 
```
void getCoordinatesKokkos(Ids, xyz, wgts) {
Ids=inputAdapter->getIdsKokkosView(); 
xyz = inputAdapter->getCoordinatesKokkosView();
if (use_weights) wgts = inputAdapter->getWeightsKokkosView();
}
```
 Method getCoordinatesKokkos() would be called by the Zoltan2 function, which would release the pointers before the application needed them again.

Reply:
I looked at that in the initial cut, however, some of the versions do a bit of computation and other stuff to get the coords.  Because of this I thought the safest was to cache with a copy.  I really didn't like this for the reasons you mentinoed above.